_rules/SC2-4-4-link-has-name.md

```diff
@@ -0,0 +1,157 @@
+
+### Expectation
+
+Each target element has an [accessible name](#accessible-name) that is [non-empty](#non-empty).
+
+## Assumptions
+
+- The rule assumes that all links are user interface components as defined by WCAG 2. When the link role is used incorrectly, this assumption may not be true.
+
+## Accessibility Support
+
+There are no major accessibility support issues known for this rule.
+
+## Background
+
+- [https://www.w3.org/WAI/WCAG20/quickref/?showtechniques=244#navigation-mechanisms-refs](https://www.w3.org/WAI/WCAG20/quickref/?showtechniques=244#navigation-mechanisms-refs)
+- [ARIA7: Using aria-labelledby for link purpose](https://www.w3.org/TR/WCAG20-TECHS/ARIA7.html)
+- [ARIA8: Using aria-label for link purpose](https://www.w3.org/TR/WCAG20-TECHS/ARIA8.html)
+- [F89: Failure of Success Criteria 2.4.4, 2.4.9 and 4.1.2 due to using null alt on an image where the image is the only content in a link](http://www.w3.org/TR/WCAG20-TECHS/F89.html)
+
+## Test Cases
+
+### Passed
+```html
+<!-- <a> element with accessible name through content -->
+<a href="http://www.w3.org/WAI"> Web Accessibility Initiative (WAI) </a>
+```
+
+```html
+<!-- element with eplicit role of link with accessible name through content -->
+<div role="link"> Web Accessibility Initiative (WAI) </div>
+```
+
+```html
+<!-- Button with the role of link -->
+<button role="link">Click me!</button>
+```
+
+```html
+<!-- Accessible name for link via aria-label -->
+<a href="http://www.w3.org/WAI"><img src="#" aria-label="This is a link"/></a>
+```
+
+```html
+<!-- Link named via title on link -->
+<a href="http://www.w3.org/WAI" title="This is a link"><img src="#" /></a>
+```
+
+```html
+<!-- Link named via title on image -->
+<a href="http://www.w3.org/WAI"><img src="#" title="This is a link"/></a>
+```
+
+```html
+<!-- Link with both image and text -->
+<a href="http://www.w3.org/WAI"><img src="#" />This is my link text</a>
+```
+
+```html
+<!-- Accessible name for link via aria-labelledby -->
+<a href="http://www.w3.org/WAI"><img src="#" aria-labelledby="id1"/></a>
+<div id="id1">This is my link text</div>
+```
+
+```html
+<!-- Passing because this rule doesn't consider whether the accessible name is descriptive enough -->
+<div>
+  <p>
+    Some detailed article, but the link to read more is placed with no relevance to the text, thereby the link has lost context.
+  </p>
+  <p>
+    <a href="readmore.html">Read More...</a>
+  </p>
+</div>
+```
+
+### Failed
+
+```html
+<!-- Image link without accessible name-->
+<a href="http://www.w3.org/WAI"><img src="#" /></a>
+```
+
+```html
+<!-- Image link where image is marked as decorative -->
+<a href="http://www.w3.org/WAI"><img src="#" alt="" /></a>
+```
+
+```html
+<!-- Link with icon inserted via font-awesome -->
+<a href="http://www.w3.org/WAI"><i class="fa fa-download"></i></a>
+```
+
+```html
+<!-- Link with image that has empty title -->
+<a href="http://www.w3.org/WAI"><img src="#" title=""/></a>
+```
+
+```html
```

Comment:
I think we're overdoing it with the img tests here.

Reply:
No harm in having multiple tests.